File: src/main/java/io/github/smootheez/curseforge/CurseforgePublisher.java
package io.github.smootheez.curseforge;

import com.google.gson.reflect.*;
import io.github.smootheez.*;
import io.github.smootheez.exception.*;
import okhttp3.*;
import org.gradle.api.*;

import java.io.*;
import java.util.*;

/**
 * Handles publishing mod files to CurseForge using the CurseForge upload API.
 * <p>
 * This publisher is responsible for:
 * <ul>
 *     <li>Resolving configured game versions, loaders, and environments</li>
 *     <li>Constructing metadata for the upload request</li>
 *     <li>Building and sending a multipart upload request to CurseForge</li>
 *     <li>Logging responses and raising exceptions for failed uploads</li>
 * </ul>
 */
public class CurseforgePublisher extends Publisher {

    /** Upload endpoint for submitting new files to a CurseForge project. */
    private static final String UPLOAD_URL = "https://minecraft.curseforge.com/api/projects/%s/upload-file";

    /** Endpoint for retrieving CurseForge's game version metadata. */
    private static final String GAME_VERSIONS_URL = "https://minecraft.curseforge.com/api/game/versions";

    /**
     * Constructs a new CurseForge publisher instance.
     *
     * @param project   the Gradle project
     * @param extension the plugin extension containing user configuration
     * @param client    the HTTP client used for API communication
     */
    public CurseforgePublisher(Project project, McModPublisherExtension extension, OkHttpClient client) {
        super(project, extension, client);
    }

    /**
     * Publishes all configured files to CurseForge.
     * <p>
     * Steps performed include:
     * <ol>
     *     <li>Reading configuration (token, project ID, load targets)</li>
     *     <li>Resolving valid CurseForge game version IDs</li>
     *     <li>Constructing upload metadata</li>
     *     <li>Sending each file to CurseForge using a multipart upload</li>
     * </ol>
     * If no valid game versions are found, the publish process is aborted.
     */
    @Override
    public void publish() {
        var curseforge = extension.getCurseforge();
        var token = curseforge.getToken().trim();
        var projectId = curseforge.getProjectId().trim();
        var iterator = extension.getFiles().getFiles().iterator();

        var desiredMcVersions = extension.getGameVersions();
        var desiredLoaders = extension.getLoaders().stream()
                .map(this::mapLoaderToCF)
                .toList();

        var desiredEnvs = curseforge.getEnvironmentType().stream()
                .map(this::mapEnvironmentToCF)
                .toList();

        var validGameVersions = fetchGameVersions(token).stream()
                .filter(tag -> {
                    var type = tag.gameVersionTypeId();
                    return type == 77784 || type == 68441 || type == 75208;
                })
                .filter(tag ->
                        desiredMcVersions.contains(tag.name()) ||
                                desiredLoaders.contains(tag.name()) ||
                                desiredEnvs.contains(tag.name())
                )
                .map(GameVersionTag::id)
                .toList();

        if (validGameVersions.isEmpty()) {
            project.getLogger().error("No valid game versions found. Please check your game versions in the configuration.");
            return;
        }

        var dependencyList = curseforge.getDependencies().stream()
                .map(dep -> ProjectsMetadata.builder()
                        .slug(dep.getSlug())
                        .relationType(dep.getRelationType())
                        .build())
                .toList();

        var metadata = curseforgeMetadata(curseforge, validGameVersions, dependencyList);
        project.getLogger().lifecycle("Curseforge metadata: " + GSON.toJson(metadata));

        project.getLogger().lifecycle("Publishing to Curseforge...");
        publishingToCurseforge(metadata, iterator, projectId, token);
    }

    /**
     * Maps internal loader types to CurseForge’s corresponding loader name.
     *
     * @param loader the loader type specified in the configuration
     * @return CurseForge’s loader label
     */
    private String mapLoaderToCF(LoaderType loader) {
        return switch (loader) {
            case FABRIC -> "Fabric";
            case QUILT -> "Quilt";
            case FORGE -> "Forge";
            case NEOFORGE -> "NeoForge";
        };
    }

    /**
     * Maps internal environment types to CurseForge’s client/server labels.
     *
     * @param env the environment type
     * @return CurseForge’s environment name
     */
    private String mapEnvironmentToCF(EnvironmentType env) {
        return switch (env) {
            case CLIENT -> "Client";
            case SERVER -> "Server";
        };
    }

    /**
     * Performs the multipart upload to CurseForge.
     *
     * @param metadata  the metadata payload describing the uploaded file
     * @param iterator  iterator over files scheduled for upload
     * @param projectId the CurseForge project ID
     * @param token     authentication token
     */
    private void publishingToCurseforge(CurseforgeMetadata metadata, Iterator<File> iterator, String projectId, String token) {
        var multipartBuilder = new MultipartBody.Builder().setType(MultipartBody.FORM);

        multipartBuilder.addFormDataPart(
                Constants.METADATA,
                null,
                RequestBody.create(GSON.toJson(metadata), MediaType.parse(Constants.MEDIA_TYPE_JSON))
        );

        var file = iterator.next();
        multipartBuilder.addFormDataPart(
                "file",
                file.getName(),
                RequestBody.create(file, MediaType.parse(Constants.MEDIA_TYPE_JAR))
        );

        var requestBody = multipartBuilder.build();
        var request = new Request.Builder()
                .url(String.format(UPLOAD_URL, projectId))
                .header("X-Api-Token", token)
                .header("User-Agent", Constants.USER_AGENT)
                .post(requestBody)
                .build();

        try (var response = client.newCall(request).execute()) {
            var body = response.body().string();

            project.getLogger().lifecycle("Response received: code=" + response.code() + " message=" + response.message());
            project.getLogger().lifecycle("Response body: " + body);

            if (!response.isSuccessful()) {
                throw new FailedFileUploadException(
                        "Failed to upload mod to Curseforge: " +
                                response.code() + " - " + response.message() + " - BODY: " + body
                );
            }

            project.getLogger().lifecycle("Successfully uploaded mod to Curseforge!");
        } catch (IOException e) {
            throw new FailedFileUploadException("Failed to upload mod to Curseforge: " + e.getMessage());
        }
    }

    /**
     * Builds a {@link CurseforgeMetadata} instance based on configuration and valid game versions.
     *
     * @param curseforge        CurseForge-specific configuration
     * @param validGameVersions resolved CurseForge game version IDs
     * @param dependencyList    resolved dependency metadata
     * @return a fully constructed CurseForge metadata object
     */
    private CurseforgeMetadata curseforgeMetadata(
            CurseforgeConfig curseforge,
            List<Integer> validGameVersions,
            List<ProjectsMetadata> dependencyList
    ) {
        var builder = CurseforgeMetadata.builder()
                .changelog(extension.getChangelog())
                .changelogType(curseforge.getChangelogType())
                .displayName(extension.getDisplayName())
                .gameVersions(validGameVersions)
                .releaseType(extension.getReleaseType())
                .isMarkedForManualRelease(curseforge.isManualRelease());

        if (!dependencyList.isEmpty()) {
            builder.relations(new Projects(dependencyList));
        }

        return builder.build();
    }

    /**
     * Fetches the full list of CurseForge game version tags.
     *
     * @param token the API token for authentication
     * @return list of available game version tags
     * @throws FailedFetchGameVersionsException if the request fails
     */
    private List<GameVersionTag> fetchGameVersions(String token) {
        var request = new Request.Builder()
                .url(GAME_VERSIONS_URL)
                .header("X-Api-Token", token)
                .header("User-Agent", Constants.USER_AGENT)
                .get()
                .build();

        try (var response = client.newCall(request).execute()) {
            var body = response.body().string();

            if (!response.isSuccessful()) {
                throw new FailedFetchGameVersionsException(
                        "Failed to fetch game versions: " +
                                response.code() + " - " + response.message() + " - BODY: " + body
                );
            }

            return GSON.fromJson(body, new TypeToken<List<GameVersionTag>>() {}.getType());
        } catch (IOException e) {
            throw new FailedFetchGameVersionsException("Failed to fetch game versions" + e.getMessage());
        }
    }
}

File: src/main/kotlin/io/github/smootheez/curseforge/CurseforgePublisher.kt
package io.github.smootheez.curseforge

import io.github.smootheez.McModPublisherExtension
import io.github.smootheez.PublisherConfig
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.gradle.api.Project
import java.io.File

// CurseForge API endpoint for uploading files
private const val UPLOAD_URL = "https://minecraft.curseforge.com/api/projects/%s/upload-file"
private const val GAME_VERSIONS_URL = "https://minecraft.curseforge.com/api/api/game/versions"

@Serializable
private data class GameVersionTag(
    val id: Int,
    val gameVersionTypeID: Int,
    val name: String,
    val slug: String
)

class CurseforgePublisher(
    private val project: Project,
    private val extension: McModPublisherExtension,
    private val client: OkHttpClient = OkHttpClient()
) {
    fun publish() {
        val curseforge = extension.curseforge
        val token = curseforge.token?.trim()
        val projectId = curseforge.projectId?.trim()
        val files = extension.files.files

        project.logger.lifecycle("🚀  Publishing to CurseForge...")

        // ✅ Validate required fields
        val missing = buildList {
            if (token.isNullOrBlank()) add("CurseForge token")
            if (projectId.isNullOrBlank()) add("Project ID")
            if (files.isEmpty()) add("File to upload (CurseForge allows only one)")
        }

        if (missing.isNotEmpty()) {
            project.logger.error("❌ Missing required configuration: ${missing.joinToString(", ")}. Aborting.")
            return
        }

        project.logger.lifecycle("✅ Found ${files.size} file(s) for upload.")
        project.logger.lifecycle("✅ All required configuration is set.")

        // Fetch allowed Minecraft game versions from CurseForge API
        project.logger.lifecycle("⏳ Fetching valid game versions from CurseForge...")
        val validVersions = fetchValidGameVersions(token!!)
        val userVersions = extension.gameVersions.map { it.trim() }
        val resolvedVersionIds = userVersions.mapNotNull { name -> validVersions.firstOrNull { it.name.equals(name, ignoreCase = true) }?.id }

        if (resolvedVersionIds.isEmpty()) {
            project.logger.error("❌ Failed to fetch valid game versions from CurseForge. Aborting.")
            return
        }
        project.logger.lifecycle("✅ Retrieved ${resolvedVersionIds.size} valid game version(s).")

        // Build metadata JSON
        project.logger.lifecycle("⏳ Preparing metadata...")

        val relations = curseforge.dependencies.asMap.values.map {
            ProjectsMetadata(
                projectID = it.projectId,
                relationType = it.relationType
            )
        }

        val metadata = CurseforgeMetadata(
            changelog = extension.changelog,
            changelogType = curseforge.changelogType,
            displayName = extension.displayName,
            gameVersions = resolvedVersionIds,
            releaseType = extension.releaseType,
            isMarkedForManualRelease = curseforge.manualRelease,
            relations = Projects(relations)
        )

        val jsonMetadata = Json.encodeToString(metadata)
        project.logger.lifecycle("📦 Metadata prepared: $jsonMetadata")

        // Construct multipart form (metadata + file)
        project.logger.lifecycle("⏳ Uploading to CurseForge...")

        publishingToCurseforge(jsonMetadata, files, projectId!!, token)
    }

    private fun fetchValidGameVersions(token: String): List<GameVersionTag> {
        return try {
            val request = Request.Builder()
                .url(GAME_VERSIONS_URL)
                .addHeader("User-Agent", PublisherConfig.USER_AGENT)
                .addHeader("X-Api-Token", token)
                .build()

            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return emptyList()
                val body = response.body.string()
                if (body.isBlank()) return emptyList()
                Json.decodeFromString(body)
            }
        } catch (e: Exception) {
            project.logger.error("❌ Failed to fetch CurseForge game versions: ${e.message}")
            emptyList()
        }
    }

    private fun publishingToCurseforge(
        jsonMetadata: String,
        files: Set<File>,
        projectId: String,
        token: String
    ) {
        val multipartBuilder = MultipartBody.Builder().setType(MultipartBody.FORM)
        multipartBuilder.addFormDataPart(
            "metadata",
            null,
            jsonMetadata.toRequestBody(PublisherConfig.MEDIA_TYPE_JSON.toMediaType())
        )

        // CurseForge accepts only a single file per upload
        val file = files.first()
        project.logger.lifecycle("📁 Attaching file: ${file.name}")

        multipartBuilder.addFormDataPart(
            "file",
            file.name,
            file.asRequestBody(PublisherConfig.MEDIA_TYPE_JAR.toMediaType())
        )

        val request = Request.Builder()
            .url(String.format(UPLOAD_URL, projectId))
            .addHeader("User-Agent", PublisherConfig.USER_AGENT)
            .addHeader("X-Api-Token", token)
            .post(multipartBuilder.build())
            .build()

        try {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    val body = response.body.string()
                    project.logger.error("❌ Upload failed: ${response.code} ${response.message}")
                    if (body.isNotBlank()) project.logger.error("Response body: $body")
                } else {
                    project.logger.lifecycle("✅ Successfully uploaded to CurseForge!")
                    val responseBody = response.body.string()
                    if (responseBody.isNotBlank()) project.logger.lifecycle("📨 Response: $responseBody")
                }
            }
        } catch (e: Exception) {
            project.logger.error("❌ Upload to CurseForge failed due to network or I/O error.")
            project.logger.error("   ➜ ${e::class.simpleName}: ${e.message}")
        }
    }
}